packages/teams-js/src/internal/communication.ts

```diff
@@ -615,39 +664,52 @@ function handleParentMessage(evt: DOMMessageEvent): void {
+    const callbackId = retrieveMessageUUIDFromResponse(message);
+    const callback = CommunicationPrivate.callbacks.get(callbackId);
+    logger('Received a response from parent for message %i', callbackId);
     if (callback) {
-      logger('Invoking the registered callback for message %i with arguments %o', message.id, message.args);
+      logger('Invoking the registered callback for message %i with arguments %o', callbackId, message.args);
       // eslint-disable-next-line @typescript-eslint/ban-ts-comment
       // @ts-ignore
       callback.apply(null, [...message.args, message.isPartialResponse]);
 
       // Remove the callback to ensure that the callback is called only once and to free up memory if response is a complete response
       if (!isPartialResponse(evt)) {
-        logger('Removing registered callback for message %i', message.id);
-        delete CommunicationPrivate.callbacks[message.id];
+        logger('Removing registered callback for message %i', callbackId);
+        CommunicationPrivate.callbacks.delete(callbackId);
+        if (!message.uuid) {
+          delete CommunicationPrivate.legacyMessageIdsToUuidMap[message.id];
+        }
       }
     }
-    const promiseCallback = CommunicationPrivate.promiseCallbacks[message.id];
+    const promiseCallback = CommunicationPrivate.promiseCallbacks.get(callbackId);
     if (promiseCallback) {
-      logger('Invoking the registered promise callback for message %i with arguments %o', message.id, message.args);
+      logger('Invoking the registered promise callback for message %i with arguments %o', callbackId, message.args);
       promiseCallback(message.args);
 
-      logger('Removing registered promise callback for message %i', message.id);
-      delete CommunicationPrivate.promiseCallbacks[message.id];
+      logger('Removing registered promise callback for message %i', callbackId);
+      CommunicationPrivate.promiseCallbacks.delete(callbackId);
+      if (!message.uuid) {
+        delete CommunicationPrivate.legacyMessageIdsToUuidMap[message.id];
+      }
     }
-    const portCallback = CommunicationPrivate.portCallbacks[message.id];
+    const portCallback = CommunicationPrivate.portCallbacks.get(callbackId);
     if (portCallback) {
-      logger('Invoking the registered port callback for message %i with arguments %o', message.id, message.args);
+      logger('Invoking the registered port callback for message %i with arguments %o', callbackId, message.args);
       let port: MessagePort | undefined;
       if (evt.ports && evt.ports[0] instanceof MessagePort) {
         port = evt.ports[0];
       }
       portCallback(port, message.args);
 
-      logger('Removing registered port callback for message %i', message.id);
-      delete CommunicationPrivate.portCallbacks[message.id];
+      logger('Removing registered port callback for message %i', callbackId);
+      CommunicationPrivate.portCallbacks.delete(callbackId);
```

Comment:
Will calls to delete run into the same problem with objects as keys? Do we need a delete function that also looks up the keys by value?

Reply:
I modified the callback above to actually return rather than the value, this way we only need to do the map lookup once and then we can both acquire and delete the values using the same key value. So I don't believe this to be an issue